### PR TITLE
Replace `user-home` with `os.homedir`.

### LINKED ISSUE #2596: Consider replacing `user-home` with `os.homedir`.
**Do you want to request a *feature* or report a *bug*?**
I don't think it counts as either. Reworking internals?

**What is the current behavior?**
The `user-home` package is used to get the user home directory.
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
`os.homedir` is used to get the user home directory.
**Please mention your node.js, yarn and operating system version.**

### PR DESCRIPTION
**Pull request by wtgtybhertgeghgtwtg**
**Summary**
Replaces `user-home` with `os.homedir`.  Proposed to close #2596.

**Test plan**
Refactors internals.  Nothing should change.

### CODE REVIEW

**Comment by bestander** [comment]
I think user-home wraps os.homedir and has some important edge case handling.
@Daniel15 has some background on it IIRC

**Comment by wtgtybhertgeghgtwtg** [comment]
`user-home` is just
```
'use strict';
module.exports = require('os-homedir')();
```
So the only edge case would be `node<4`.

**Comment by bestander** [comment]
Lol, indeed it is https://github.com/sindresorhus/user-home/blob/master/index.js.

File: src/registries/npm-registry.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ import {addSuffix, removePrefix} from '../util/misc';
 import isRequestToRegistry from './is-request-to-registry.js';
 
 const defaults = require('defaults');
-const userHome = require('user-home');
+const userHome = require('os').homedir();
```
Comment (bestander) [inline src/registries/npm-registry.js]: Considering that constants.js has custom edge case handling to homedir, can you extract that part into an internal module in `src/util`?
I think that edge case handling may cause some issues

Comment (wtgtybhertgeghgtwtg) [inline src/registries/npm-registry.js]: Would you like me to use that internal module for all uses of `os.homedir`?  What would you like me to name that module?

Comment (bestander) [inline src/registries/npm-registry.js]: Yeah, I think we should use one homedir definition everywhere.
Use your best judgement and lowercase letters, something with words user, home and dir maybe :)

Comment (wtgtybhertgeghgtwtg) [inline src/registries/npm-registry.js]: Being up at 4 AM to fix PR's is the highlight of my day, so I wouldn't bank on my best judgement.
It needs access to `isRootUser` (which is exported) and `getUid` (which is not).  Would you like me to make `getUid` an export of `src/constants.js`, or move both to `src/util`?

**Comment by wtgtybhertgeghgtwtg** [comment]
Tested failing because `babel-loader` [deleted the branch](https://github.com/babel/babel-loader/pull/353) [that was being checked](https://travis-ci.org/yarnpkg/yarn/jobs/199455023#L736).
Why does it check a branch of `babel-loader`, anyway?

**Comment by bestander** [comment]
Thanks, @wtgtybhertgeghgtwtg.
I'll fix the test now.
We just needed a package to test that resolver works for this type of dependency.

**Comment by bestander** [comment]
This should be safe to merge, all the other tests pass.

File: src/util/user-home-dir.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/* @flow */
+const path = require('path');
+const {ROOT_USER} = require('../constants');
```
Comment (connesc) [inline src/util/user-home-dir.js]: `ROOT_USER` is undefined at this point due to the cyclic dependency.

Comment (Daniel15) [inline src/util/user-home-dir.js]: Good catch, that explains #2807.

Comment (bestander) [inline src/util/user-home-dir.js]: @connesc would you send a PR then?

Comment (connesc) [inline src/util/user-home-dir.js]: Sure: #2855 :wink:
